### ISSUE TITLE
Python Black format_on_save doesn't work with Nushell as login Shell

### ISSUE THREAD

**Issue by Tom-Boscher** [zed-industries/community#2106]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Using Nushell as a default shell, Black `format_on_save` doesn't work anymore.
Which argument should I use for Black to use it in Zed ?
Using Zsh i had:
```
"Python": {
  "format_on_save": "on",
  "formatter": {
    "external": {
      "command": "black",
      "arguments": [
        "-" // stdin
      ]
    }
  }
}
```

```
2023-09-30T20:21:54 [ERROR] failed to format via external command "black"

Caused by:
    No such file or directory (os error 2)
```

Black is installed in the venv and globally, I can format my file in my Shell using Nushell without any issue, just the `format_on_save` doesn't work.

### Environment

Zed: v0.106.2 (preview)
OS: macOS 14.0.0
Memory: 16 GiB
Architecture: aarch64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

_No response_

**Comment by krstp** [comment]
Indeed! Thanks for posting this! It has driven me crazy trying to figure it out on multiple syntaxes that worked before, but are not working now, the core being the need to provide the full path. Big help. Thanks.

**Comment by kingwingfly** [comment]
I'm also using a Mac, and it is certainly that a software must use full path to call a binary instead name in env path.

However, zed on my machine just clear all the content in my python file when I use format on save.

**Comment by krstp** [comment]
Posting here as this is somewhat related: https://github.com/zed-industries/community/issues/1380#issuecomment-1819173317

**Comment by krstp** [comment]
On another note, creating `pyrightconfig.json` in the project folder helps immensely, i.e.:
```
{
  "venvPath": "/Users/<user>/.pyenv/versions",
  "venv": "core"
}
```
where:

- `venvPath` is the path to your virtual environment, this way we can have multiple Zed instances (projects) running on different virtuanenv..
- `venv` is the name of your virtual environment

In above case I work with `pyenv`.

**Comment by krstp** [comment]
@kingwingfly One more note to help you out. Check installation of your `black`, then make sure it is correctly referenced in the `command` field. To dissect it more, you can install `black` or any other tool two ways:
- have it assigned `globally`, this is the case when you install it using `brew`
     - this is the command by default that will be used by ZED, whether `ruff` or `black` or anything else, `brew` will yield the global definition, i.e. `"command": "black",`
- have it assigned `locally` per virtual environment.
    - Personally, I am the user of `pyenv`, so this is how I am using it, but this approach should generally work with any other virtual env management. To the point, if you want to use `local` (in-virtual env) package distribution provide the path to your virtual env installed package, i.e. where all your virt env `shims` (packages) are located: `"command": "/Users/<user>/.pyenv/shims/black",`
 
PS. AFAIK `shims` became a versatile across diff venv management solutions (`pipenv`, etc.), this is what I've seen, but I might be mistaken now, my memory span tends to be short on trivial things :)

**Comment by kingwingfly** [comment]
@krstp Thanks for your reply! But I just give up. 😎
I tried `conda`, `brew`, `venv`, modified `pyrightconfig.json`, and config relative and absolute path, call `sh` script, but all of them just not work, or even just clean all the text buffer of the file in current tab.

Luckily, I do not write python frequently, and I do test in docker, so I just call black in Dockerfile CMD. 

I remember, years ago, `black` does not work on vscode for several months either, now they have a plugin for `black`. Probably it's just `black`'s problem.

**Comment by krstp** [comment]
Well last thing I can advise, make sure your  `settings.json` is correctly formatted: no missing commas, make sure end attribute does not close with comma, etc. Good ad-hoc way to verify your settings is read correctly is to change font... if `json` is not formatted correctly, ZED will default to its default font.

Otherwise, launch terminal `Ctrl+`` `  or `CMD+J` invoke python (default or the one from your env path) and make sure you can execute `black`. If not, it should give you additional pointers.
